### ISSUE TITLE
Possible to inherit from the RequiredIf attribute?

### ISSUE THREAD

**Issue by konstrui** [JaroslawWaliszko/ExpressiveAnnotations#70]
I want to know if I can inherit from the RequiredIfAttribute and override the FormatMessage method. I need to create a 'localized' attribute so that I can get translations for the error messages.

**Comment by JaroslawWaliszko** [comment]
Hi,
The `sealed` keyword is there by design. I've introduced it I after going through several sources, e.g.:

* [MSDN](http://msdn.microsoft.com/en-us/library/ms182267(v=VS.100).aspx):

 >CA1813: Avoid unsealed attributes: The .NET Framework class library provides methods for retrieving custom attributes. By default, these methods search the attribute inheritance hierarchy; for example Attribute.GetCustomAttribute searches for the specified attribute type, or any attribute type that extends the specified attribute type. Sealing the attribute eliminates the search through the inheritance hierarchy, and can improve performance.

* [Framework Design Guidelines](http://www.amazon.com/dp/0321545613):

 > DO seal custom attribute classes, if possible. This makes the look-up for the attribute faster.

* [Jon Skeet note](http://stackoverflow.com/a/268287/270315):

 >Classes should either be designed for inheritance or prohibit it. There is a cost to designing for inheritance:
 >* It can pin down your implementation (you have to declare which methods are going to call which other methods, in case a user overrides one but not the other)
 >* It reveals your implementation rather than just the effects
 >* It means you have to think of more possibilities when designing
 >* Things like Equals are hard to design in an inheritance tree
 >* It requires more documentation
 >* An immutable type which is subclassed may become mutable (ick)
 >
 ><br />Item 17 of Effective Java goes into more details on this - regardless of the fact that it's written in the context of Java, the advice applies to .NET as well.
 >
 >Personally I wish classes were sealed by default in .NET.

To summarize, I wasn't considering the inheritance there and can't say right now it is a correct approach. Probably the better idea is to inherit from abstract `ExpressiveAttribute` class.

**Comment by konstrui** [comment]
I see. I will try that. Thanks!

--
Dit bericht is verstuurd vanaf mijn telefoon en is om die reden kort en direct.

Op 4 aug. 2015 om 22:38 heeft Jaroslaw Waliszko <notifications@github.com<mailto:notifications@github.com>> het volgende geschreven:


Hi,
The sealed keyword is there by design. I've introduced it I after going through several sources, e.g.:

  *   MSDN<http://msdn.microsoft.com/en-us/library/ms182267(v=VS.100).aspx>:

CA1813: Avoid unsealed attributes: The .NET Framework class library provides methods for retrieving custom attributes. By default, these methods search the attribute inheritance hierarchy; for example Attribute.GetCustomAttribute searches for the specified attribute type, or any attribute type that extends the specified attribute type. Sealing the attribute eliminates the search through the inheritance hierarchy, and can improve performance.

  *   Framework Design Guidelines<http://www.amazon.com/dp/0321545613>:

DO seal custom attribute classes, if possible. This makes the look-up for the attribute faster.

  *   Jon Skeet note<http://stackoverflow.com/a/268287/270315>:

Classes should either be designed for inheritance or prohibit it. There is a cost to designing for inheritance:

     *   It can pin down your implementation (you have to declare which methods are going to call which other methods, in case a user overrides one but not the other)
     *   It reveals your implementation rather than just the effects
     *   It means you have to think of more possibilities when designing
     *   Things like Equals are hard to design in an inheritance tree
     *   It requires more documentation
     *   An immutable type which is subclassed may become mutable (ick)

Item 17 of Effective Java goes into more details on this - regardless of the fact that it's written in the context of Java, the advice applies to .NET as well.

Personally I wish classes were sealed by default in .NET.

To summarize, I wasn't considering the inheritance there and can't say right now it is a correct approach. Probably the better idea is to inherit from abstract ExpressiveAttribute class.

-

**Comment by JaroslawWaliszko** [comment]
No problem. I'd just like to add that if localization is the main issue, take look at the [sample project](https://github.com/JaroslawWaliszko/ExpressiveAnnotations/tree/master/src/ExpressiveAnnotations.MvcWebSample), where the error messages are localized. Maybe that solution will be sufficient for your purposes. 

Regards.

**Comment by JaroslawWaliszko** [comment]
An example of inheritance from `ExpressiveAttribute` is shown below:
```
public class AssertThat2Attribute : ExpressiveAttribute
{
    public AssertThat2Attribute(string expression)
        : base(expression, "redefined default message for {0} field with {1} logic")
    {
    }

    protected override ValidationResult IsValidInternal(object value, ValidationContext validationContext)
    {
        if (value != null)
        {
            Compile(validationContext.ObjectType);
            if (!CachedValidationFuncs[validationContext.ObjectType](validationContext.ObjectInstance))
                return new ValidationResult(
                    FormatErrorMessage(validationContext.DisplayName, Expression),
                    new[] { validationContext.MemberName });
        }

        return ValidationResult.Success;
    }
}
```
Next, if client-side validation is needed as well, new validator needs to be provided for such a new attribute to be recognized:
```
public class AssertThat2Validator : ExpressiveValidator<AssertThat2Attribute>
{
    public AssertThat2Validator(ModelMetadata metadata, ControllerContext context, AssertThat2Attribute attribute) 
        : base(metadata, context, attribute)
    {
    }

    public override IEnumerable<ModelClientValidationRule> GetClientValidationRules()
    {
        var rule = GetBasicRule("assertthat");
        yield return rule;
    }
}
```
Global.asax wire-up can be done in the following manner:
```
protected void Application_Start()
{
    DataAnnotationsModelValidatorProvider.RegisterAdapter(
        typeof(AssertThat2Attribute), typeof(AssertThat2Validator));
    ...
```
If the [`Priority`](https://github.com/JaroslawWaliszko/ExpressiveAnnotations#how-to-construct-conditional-validation-attributes) property has to be recognized, different Global.asax setup is required, namely:
```
protected void Application_Start()
{
    ModelValidatorProviders.Providers.Remove(
        ModelValidatorProviders.Providers.FirstOrDefault(x => x is DataAnnotationsModelValidatorProvider));
    ModelValidatorProviders.Providers.Add(new ExpressiveAnnotationsModelValidatorProvider2());
    ...
```
As you can see, custom version of `ExpressiveAnnotationsModelValidatorProvider` is used. It differs from default one, because recognizes newly defined attribute. Its implementation is as straightforward as that:
```
public class ExpressiveAnnotationsModelValidatorProvider2 :  ExpressiveAnnotationsModelValidatorProvider
{
    public ExpressiveAnnotationsModelValidatorProvider2()
    {
        RegisterAdapter(typeof(AssertThat2Attribute), typeof(AssertThat2Validator));
        ...
    }
}
```

**Comment by senzacionale** [comment]
Just for info. Code works before, but after new version I can not use translation message any more. Now is always in english. Why? What has changed?

**Comment by jwaliszko** [comment]
@senzacionale: I haven't noticed that, could you show any reproducible example?

**Comment by senzacionale** [comment]
ErrorMessage code:
private const string _defaultErrorMessage = "The {0} field is required by the following logic: {1}";

public override IEnumerable<ModelClientValidationRule> GetClientValidationRules()
        {
            var rule = GetBasicRule("requiredif");
            rule.ValidationParameters.Add("allowempty", AllowEmpty.ToJson());
            yield return rule;
        }

rule.ErrorMessage is "The SchoolName field is required by the following logic: IsCustomerInputRequired == true"

but in UI I get back 

The SchoolName field is required.

I checked jquery.validate.unobtrusive.js file

function setValidationValues(options, ruleName, value)  method has this message and not message from ExpressiveAnnotations

**Comment by senzacionale** [comment]
Thank you for reply. I copy code in github

regards,
Mitja

2015-12-10 22:19 GMT+01:00 Jarosław Waliszko <notifications@github.com>:

> @senzacionale <https://github.com/senzacionale>: I haven't noticed that,
> could you show any reproducible example?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jwaliszko/ExpressiveAnnotations/issues/70#issuecomment-163752300>
> .
>



-- 
L.P.
Mitja Bombač

**Comment by jwaliszko** [comment]
@senzacionale: I don't understand what has happened. I need more comprehensive explanation what is wrong, ideally tiny project where it can be reproduced. What's more - I've just pushed a commit: https://github.com/jwaliszko/ExpressiveAnnotations/commit/b7127e44b86ad450af8dec3a711c98ff00dc0c73 where I've shown how to override default messages globally (if it is the problem you're talking about), and it seems to be working as expected.

**Comment by senzacionale** [comment]
I will try to prepare small project if I can reproduce it there, Only probme is that jquery validation message is not the same for "RequiredIf". Problem is only with jquery error message. So I don't know if is jquery and MVC5 or something else now.

**Comment by jwaliszko** [comment]
@senzacionale: hi, any more update on this? Have you been able to reproduce it?

**Comment by senzacionale** [comment]
Hi, I got this error now when I try your solution:

Message=Either ErrorMessageString or ErrorMessageResourceName must be set,
but not both.
Source=System.ComponentModel.DataAnnotations

2015-12-14 13:09 GMT+01:00 Jarosław Waliszko <notifications@github.com>:

> @senzacionale <https://github.com/senzacionale>: hi, any more update on
> this? Have you been able to reproduce it?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jwaliszko/ExpressiveAnnotations/issues/70#issuecomment-164421424>
> .
>



-- 
L.P.
Mitja Bombač

**Comment by konstrui** [comment]
Just wanted to check in with a question about the override of the RequiredIfAttribute. Currently I use the following code as my custom RequiredIf attribute:

    public class LocalizedRequiredIfAttribute : RequiredIfAttribute`
    {
        private string taal;
        private string sectie;
        private string fieldName;

        [Inject]
        public IChannelFactory<IBLBgsServiceChannel> Factory { get; set; }

        public LocalizedRequiredIfAttribute(string dependentProperty, object targetValue, string sectie = null, [CallerMemberName] string fieldName = null)
            : base(dependentProperty, targetValue)
        {
            this.taal = Constants.StandaardLandcode;
            this.sectie = sectie ?? ResourceBO.DefaultSectie;
            this.fieldName = fieldName ?? "Veld";
            ErrorMessageResourceName = this.fieldName + "VerplichtOmdatAnderVeldTrueIs";
        }

        public override string FormatErrorMessage(string name)
        {
            return TaskHelper.RunSync(() => new ResourceBO(Factory).GetStringAsync(taal, sectie, ErrorMessageResourceName, name));
        }
    }


How would I transfer this to the code you mentioned above (the AssertThat2Attribute)?

**Comment by jwaliszko** [comment]
Both examples are shown here: [`CustomizedAssertThatAttribute`](https://github.com/jwaliszko/ExpressiveAnnotations/blob/adc4c9ed72f6da26607ac6e372529ae5d4dea3f4/src/ExpressiveAnnotations.MvcWebSample/Inheritance/CustomizedAssertThatAttribute.cs) and [`CustomizedRequiredIfAttribute`](https://github.com/jwaliszko/ExpressiveAnnotations/blob/adc4c9ed72f6da26607ac6e372529ae5d4dea3f4/src/ExpressiveAnnotations.MvcWebSample/Inheritance/CustomizedRequiredIfAttribute.cs).

**Comment by konstrui** [comment]
Thanks.
Do I need to use the Resources type as you use in the example? I do not make use of .NET Resources (.resx files). So I'm not sure how to implement otherwise.

As you can see I get the resource from my BusinessLayer.

**Comment by jwaliszko** [comment]
These are examples of possible modifications to default implementation. You are not forced to use the resources. As you can see [here](https://github.com/jwaliszko/ExpressiveAnnotations/blob/adc4c9ed72f6da26607ac6e372529ae5d4dea3f4/src/ExpressiveAnnotations.MvcWebSample/Inheritance/CustomizedAssertThatAttribute.cs#L22), error formatting function is called when validation fails. You can either override that method, do not use it, write your own, etc.

I believe your approach will work, at least to the extent you are going to use it - give it a try. Just a side note - be aware that when you decide to override mentioned error formatting method, you can loose the possibility of using `{field}` and `{field:n}` formatters (see [here])(https://github.com/jwaliszko/ExpressiveAnnotations#how-to-fetch-field-value-or-display-name-in-error-message) - but again, unless you provide that code.

**Comment by konstrui** [comment]
OK, got it to work. Just having a problem with using the Priority.
If I use the approach with a custom ExpressiveAnnotationsModelValidatorProvider then I get this error:

    Validation type names in unobtrusive client validation rules must be unique. The following validation type was seen more than once: required

This occurs when this code is executed:

    sb.Append(InputExtensions.RadioButtonFor(htmlHelper, expression, value, routeValueDictionary));

Any idea why this is a problem?

**Comment by jwaliszko** [comment]
It's MVC generated exception, saying that you cannot send to the client-side duplicated validation type names, i.e. rules with the same name. Your error is surprising for me, and related to MVC built-in `required` attribute, which is not even processed by EA. It's because, when you need to apply more than one attribute of the same type to a single field, EA takes care of that by diversifying the names - see [logic responsible for that](https://github.com/jwaliszko/ExpressiveAnnotations/blob/adc4c9ed72f6da26607ac6e372529ae5d4dea3f4/src/ExpressiveAnnotations.MvcUnobtrusive/Validators/ExpressiveValidator.cs#L140).

I assume, you haven't removed any existing validation providers prior to registering your own and there is a conflict. Probably you should disable existing provider - as shown few posts above, in Global.asax, try to invoke:
```
ModelValidatorProviders.Providers.Remove(
    ModelValidatorProviders.Providers.FirstOrDefault(x => x is DataAnnotationsModelValidatorProvider));
```
before:
```
ModelValidatorProviders.Providers.Add(new YourCustomModelValidatorProvider());
```

**Comment by konstrui** [comment]
Hmm... strange. I actually did that:

    // Add new ModelValidatorProvider (Expressive Annotation RequiredIf attribute)
            ModelValidatorProviders.Providers.Remove(ModelValidatorProviders.Providers.FirstOrDefault(x => x is DataAnnotationsModelValidatorProvider));
            ModelValidatorProviders.Providers.Add(new LocalizedExpressiveAnnotationsModelValidatorProvider());

Not sure. I do have some old validators that register Adapters inside there constructors:

    DataAnnotationsModelValidatorProvider.RegisterAdapter(typeof(LocalizedRangeAttribute), typeof(RangeAttributeAdapter));

**Comment by jwaliszko** [comment]
If some of them inherit at some point from `System.Web.Mvc.DataAnnotationsModelValidatorProvider`, it can be the reason. Move all of the other adapters registrations this single EA provider and it should be fine.

**Comment by konstrui** [comment]
Well, this is not working out for me :(
I've add all Adapters to my custom class as you suggested, but no luck.

I can see that e.g. another attribute (LocalizedRequiredAttribute) is inheriting from:

    public class LocalizedRequiredAttribute : System.ComponentModel.DataAnnotations.RequiredAttribute

Maybe that is a problem?

If I register my LocalizedExpressiveAnnotationsModelValidatorProvider inside the Application_Start I can see that there are allready 2 ModelValidator providers, but I don't think that is a problem though since they are not inheriting from DataAnnotationsModelValidatorProvider:

- System.Web.Mvc.DataErrorInfoModelValidatorProvider
- System.Web.Mvc.ClientDataTypeModelValidatorProvider

Out of ideas now.

**Comment by jwaliszko** [comment]
These both left are OK. This leaves me out of ideas also. Probably I can help if I get a tiny zip archive from you. Remove everything, leave just most basic stuff, instead of resources return hardcoded strings, simplify view, simplify model etc. Expose the problem only.

**Comment by konstrui** [comment]
If I use a new application and implement my own attribute with the registration as above, it works. When I add to RequiredIf attributes on the same property I get the same error with this test:

     public class TestModel
    {
        public bool IsRequired { get; set; }

        [LocalizedRequiredIf("IsRequired == false")]
        [RequiredIf("IsRequired == false")]
        public string Name { get; set; }

    }

I assumed that since my LocalizedRequiredIf inherits from the EA RequiredIf it should be no problem?

**Comment by jwaliszko** [comment]
You should inherit from `ExpressiveAttribute`. You cannot inherit from `RequiredIf` unless you're using your custom build of EA. There is a `sealed` keyword, about which this entire topic is.

**Comment by konstrui** [comment]
Ah yes! That is what I'm doing of course. :)

In my test case the **LocalizedRequiredIf** is inheriting from the **ExpressiveAttribute** and the **RequiredIf** attribute is the default one, from your package.

So, is it expected behavior that I get that error message by just using the EA attributes (one default and a custom one?) If I use two attributes of the same type, it isn't a problem.

**Comment by jwaliszko** [comment]
Current implementation doesn't allow you to mix your custom implementation of `RequiredIf\AssertThat` attributes, with original ones - but only when applying to a single field. You need to decide which one you prefer;], Across multiple fields, such annotations can be mixed.

There are technical reasons behind that. Maybe it could be redesigned, but it's not that straightforward.

Why? Briefly - it's because your custom attributes are different types than original attributes, while they are registered under the same validation types names (rules names) for client-side ([custom](https://github.com/jwaliszko/ExpressiveAnnotations/blob/adc4c9ed72f6da26607ac6e372529ae5d4dea3f4/src/ExpressiveAnnotations.MvcWebSample/Inheritance/CustomizedAssertThatValidator.cs#L16) vs [orig](https://github.com/jwaliszko/ExpressiveAnnotations/blob/adc4c9ed72f6da26607ac6e372529ae5d4dea3f4/src/ExpressiveAnnotations.MvcUnobtrusive/Validators/AssertThatValidator.cs#L36)). At the client-side we have hardcoded 2 adapters - `requiredif` and `assertthat`, so these are the only names, under which your custom attributes can be registered and later matched. Because identical rules throws exception, suffix allocation logic tries to differentiate them. Lookup fails because types are different, rules stays unchanged (suffix not generated - [check](https://github.com/jwaliszko/ExpressiveAnnotations/blob/adc4c9ed72f6da26607ac6e372529ae5d4dea3f4/src/ExpressiveAnnotations.MvcUnobtrusive/Validators/ExpressiveValidator.cs#L175)) and error is risen.

**Comment by konstrui** [comment]
Using the ExpressiveAnnotationsModelValidatorProvider approach just doesn't work for me. I have other custom attributes that for sure exist ones, but are also generating the same error when using a custom ExpressiveAnnotationsModelValidatorProvider. Very weird...

Also, is this right?

    protected override ValidationResult IsValidInternal(object value, ValidationContext validationContext)
        {
            var isEmpty = value is string && string.IsNullOrWhiteSpace((string) value);
            if (value == null || (isEmpty && !AllowEmptyStrings))
            {
                Compile(validationContext.ObjectType);
                if (CachedValidationFuncs[validationContext.ObjectType](validationContext.ObjectInstance)) // check if the requirement condition is satisfied
                    return new ValidationResult( // requirement confirmed => notify
                        FormatErrorMessage(validationContext.DisplayName, Expression, validationContext.ObjectInstance),
                        new[] {validationContext.MemberName});
            }

            return ValidationResult.Success;
        }

This isn't validating empty strings or the expression? The CachedValidationFuncs is working as I would expect.
I've included a test project.


[ValidationTest.zip](https://github.com/jwaliszko/ExpressiveAnnotations/files/177577/ValidationTest.zip)

**Comment by konstrui** [comment]
If I change my code like this, it seems to work:

    private static void ConfigureDataAnnotations()
        {
            // Do not enable default data annotations
            DataAnnotationsModelValidatorProvider.AddImplicitRequiredAttributeForValueTypes = false;

            // Register the LocalizedRangeAttribute
            DataAnnotationsModelValidatorProvider.RegisterAdapter(typeof(LocalizedRangeAttribute), typeof(RangeAttributeAdapter));

            // Register the LocalizedRequiredAttribute
            DataAnnotationsModelValidatorProvider.RegisterAdapter(typeof(LocalizedRequiredAttribute), typeof(RequiredAttributeAdapter));

            // Register the LocalizedStringLengthAttribute
            DataAnnotationsModelValidatorProvider.RegisterAdapter(typeof(LocalizedStringLengthAttribute), typeof(StringLengthAttributeAdapter));

            // Register the LocalizedRequiredIfAttribute
            DataAnnotationsModelValidatorProvider.RegisterAdapter(typeof(LocalizedRequiredIfAttribute), typeof(LocalizedRequiredIfValidator));
        }

Whenever I create a new ExpressiveAnnotationsModelValidatorProvider or even a DataAnnotationsModelValidatorProvider then all my existing custom Validation Attributes are not working anymore. Seems that they occur more than once at the moment I use the custom provider(s).

That's pitty, since I now can't make use of the Priority.

**Comment by konstrui** [comment]
If I use only server-side validation I can get it to work with the above solution. When I also 'enable' the unobtrusive clientside validation, the error message is the one from the constructor:

    public LocalizedRequiredIfAttribute(string expression, string sectie = null, string suffix = null, [CallerMemberName] string fieldName = null)
            : base(expression, "Errormessage")
        {
            _taal = Constants.StandaardLandcode;
            _sectie = sectie;
            _fieldName = fieldName ?? "Veld";
            _suffix = "_" + suffix;

            AllowEmptyStrings = false;
        }

Where I would love to see the same one has I get with server side validation:

    /// <summary>
        /// Format the error message
        /// </summary>
        /// <param name="name"></param>
        /// <returns></returns>
        public override string FormatErrorMessage(string name)
        {
            return TaskHelper.RunSync(() => new ResourceBO(Factory).GetStringWithExpressionsAsync(
                _taal,
                _sectie,
                string.Concat(_fieldName, _suffix),
                Expression,
                name));
        }

        /// <summary>
        /// Check if the value is valid
        /// </summary>
        /// <param name="value"></param>
        /// <param name="validationContext"></param>
        /// <returns></returns>
        protected override ValidationResult IsValidInternal(object value, ValidationContext validationContext)
        {
            var isEmpty = value is string && string.IsNullOrWhiteSpace((string)value);
            if (value == null || (isEmpty && !AllowEmptyStrings))
            {
                Compile(validationContext.ObjectType);

                // Check if the requirement condition is satisfied
                if (CachedValidationFuncs[validationContext.ObjectType](validationContext.ObjectInstance))

                    // Set errormessage
                    ErrorMessage = FormatErrorMessage(validationContext.DisplayName);

                    // Requirement confirmed => notify
                    return new ValidationResult(ErrorMessage, new[] { validationContext.MemberName });
            }

            return ValidationResult.Success;
        }

Is this possible?

**Comment by jwaliszko** [comment]
I've pushed a commit which should simplify that: https://github.com/jwaliszko/ExpressiveAnnotations/commit/9721b5a29d28dc7feaa0e08a34d318659c716148. Take a look at the custom attributes implementations in web sample.

Your sample is fixed as well, but GitHub doesn't allow me to attach it here, they have some bug right now. I'll attach it when it will be fixed.

**Comment by konstrui** [comment]
Ah... that is a really great addition! Thanks for that.
I'm going to test this in the following days.

**Comment by jwaliszko** [comment]
Fixed sample: [ValidationTest_FIX.zip](https://github.com/jwaliszko/ExpressiveAnnotations/files/183264/ValidationTest_FIX.zip). I've added comments after **ToDo-fix:** prefix. I've also attached DLLs, which are compiled with the https://github.com/jwaliszko/ExpressiveAnnotations/commit/9721b5a29d28dc7feaa0e08a34d318659c716148 changeset (to be shipped with next release).

**Comment by konstrui** [comment]
Thank you for the project. Very helpful information!
I think I've got it working except for the client side validation that gives me the error of duplicate rules.

The strange thing is that there error also occurs for validators other than the RequiredIf validator.

As if the DataAnnotations are added twice, but I did removed them first.
Anyway, thanks for the help!

**Comment by jwaliszko** [comment]
I believe that the error mentioned above, related to duplicated client validation rules, should be fixed now (as reported here: https://github.com/jwaliszko/ExpressiveAnnotations/issues/113).
